torch/onnx/_internal/exporter/_registration.py

```diff
@@ -209,32 +117,36 @@ def register_op(
             function: The onnx-script function to register.
             is_complex: Whether the function is a function that handles complex valued inputs.
         """
-        if not hasattr(function, "signature"):
-            try:
-                # TODO(justinchuby): Use the op_signature attribute when onnxscript is updated in CI
-                if isinstance(function, onnxscript.OnnxFunction):
-                    function.signature = _schemas.OpSignature.from_function(  # type: ignore[attr-defined]
-                        function,
-                        function.function_ir.domain,
-                        function.name,
-                        opset_version=function.opset.version,
-                    )
-                else:
-                    function.signature = _schemas.OpSignature.from_function(  # type: ignore[attr-defined]
-                        function, "__custom", function.__name__
-                    )
-            except Exception:
-                logger.exception(
-                    "Failed to infer the signature for function '%s'", function
+        signature: _schemas.OpSignature | None
+        try:
+            # TODO(justinchuby): Use the op_signature attribute when onnxscript is updated in CI
+            if isinstance(function, onnxscript.OnnxFunction):
+                signature = _schemas.OpSignature.from_function(  # type: ignore[attr-defined]
+                    function,
+                    function.function_ir.domain,
+                    function.name,
+                    opset_version=function.opset.version,
                 )
-
-        onnx_decomposition = OnnxDecompMeta(
-            onnx_function=function,
-            fx_target=target,
-            is_custom=True,
-            is_complex=is_complex,
+            else:
+                signature = _schemas.OpSignature.from_function(  # type: ignore[attr-defined]
+                    function, "__custom", function.__name__
+                )
+        except Exception:
```

Comment:
What are the possible reason it fails?

Reply:
When the function is targeting an HOP, for example, it will accept higher ordered functions as arguments and fail to generate an ONNX signature. In this case we set signature to None and dispatch to this function always.

Added as a comment